lineoperations/src/linefunctions.c

```diff
@@ -0,0 +1,401 @@
+	lines         = g_malloc(sizeof(gchar *) * totalnumlines);
+	newfile       = g_malloc(sizeof(gchar) * (totalnumchars+1));
+	numlines      = 0;
+	nfposn        = 0;
+	toremove      = NULL;
+
+	if(newfile && lines)    // verify memory allocation
+	{
+		// copy *all* lines into **lines array
+		for(i = 0; i < totalnumlines; i++)
+			lines[numlines++] = sci_get_line(doc->editor->sci, i);
+
+		// allocate and set *toremove to all TRUE
+		toremove = g_malloc(sizeof(gboolean) * numlines);
+		for(i = 0; i < (numlines); i++)
+			toremove[i] = TRUE;
+
+		// set all unique rows to FALSE
+		// toremove[i] corresponds to lines[i]
+		for(i = 0; i < numlines; i++)
+			if(toremove[i])
+				for(j = (i+1); j < numlines; j++)
+					if(toremove[j])
+						if(strcmp(lines[i], lines[j]) == 0)
+						{
+							toremove[i] = FALSE;
+							toremove[j] = FALSE;
+						}
+
+
+		// copy line into 'newfile' if it is not FALSE(unique)
+		for(i = 0; i < numlines; i++)
+		{
+			if(!toremove[i])
+				for(j = 0; lines[i][j] != '\0'; j++)
+					newfile[nfposn++] = lines[i][j];
+			if(lines[i]) g_free(lines[i]);
+		}
+
+		newfile[nfposn] = '\0';
+		sci_set_text(doc->editor->sci, newfile);	// set new document
+	}
+
+
+	// each line is freed in above for-loop
+	if(lines)    g_free(lines);
+	if(newfile)  g_free(newfile);
+	if(toremove) g_free(toremove);
+}
+
+
+
+void rmemtyln(GeanyDocument *doc) {
+	gint  totalnumchars;    // number of characters in the document
+	gint  totalnumlines;    // number of lines in the document
+	gchar *line;            // temporary line
+	gint  linelen;          // length of *line
+	gchar *newfile;         // *final* string to replace current document
+	gint  nfposn;           // position to the last character in newfile
+	gint  i;                // temporary iterator number
+	gint j;                 // iterator
+
+
+	totalnumchars = sci_get_length(doc->editor->sci);
+	totalnumlines = sci_get_line_count(doc->editor->sci);
+	/*
+	 * Allocate space for the new document (same amount as open document.
+	 * If the document contains lines with only whitespace,
+	 * not all of this space will be used.)
+	*/
+	newfile   = g_malloc(sizeof(gchar) * (totalnumchars+1));
+	nfposn    = 0;
+
+	if(newfile)    // verify memory allocation
+	{
+		for(i = 0; i < totalnumlines; i++)    // loop through opened doc char by char
+		{
+			linelen = sci_get_line_length(doc->editor->sci, i);
+
+			if(linelen == 2)
+			{
+				line = sci_get_line(doc->editor->sci, i);
+
+				if(line[0] != '\r')
+					// copy current line into *newfile
+					for(j = 0; line[j] != '\0'; j++)
+						newfile[nfposn++] = line[j];
+			}
+			else if(linelen != 1)
+			{
+				line = sci_get_line(doc->editor->sci, i);
+
+				// copy current line into *newfile
+				for(j = 0; line[j] != '\0'; j++)
+					newfile[nfposn++] = line[j];
+			}
+		}
+		newfile[nfposn] = '\0';
+		sci_set_text(doc->editor->sci, newfile);	// set new document
+
+
+		g_free(newfile);
+	}
```

Comment:
I'd probably rather implement this one using only direct Scintilla manipulation:
```C
gint line_count;

sci_start_undo_action(sci);

line_count = sci_get_line_count(sci);
for (gint line = 0; line < line_count; line++) {
  if (sci_get_position_from_line(sci, line) == sci_get_line_end_position(sci, line)) {
    scintilla_send_message(sci, SCI_DELETERANGE,
                           sci_get_position_from_line(sci, line),
                           sci_get_line_length(sci, line));
    line_count--;
  }
}

sci_end_undo_action(sci);
```
or similar